src/table_builder.cc

```diff
@@ -126,36 +163,63 @@ void TitanTableBuilder::AddBlob(const Slice& key, const Slice& value,
         new BlobFileBuilder(db_options_, cf_options_, blob_handle_->GetFile()));
   }
 
-  RecordTick(statistics(stats_), TITAN_BLOB_FILE_NUM_KEYS_WRITTEN);
-  RecordInHistogram(statistics(stats_), TITAN_KEY_SIZE, key.size());
-  RecordInHistogram(statistics(stats_), TITAN_VALUE_SIZE, value.size());
-  AddStats(stats_, cf_id_, TitanInternalStats::LIVE_BLOB_SIZE, value.size());
-  bytes_written_ += key.size() + value.size();
+  RecordInHistogram(statistics(stats_), TITAN_KEY_SIZE, record.key.size());
+  RecordInHistogram(statistics(stats_), TITAN_VALUE_SIZE, record.value.size());
+  AddStats(stats_, cf_id_, TitanInternalStats::LIVE_BLOB_SIZE,
+           record.value.size());
+  bytes_written_ += record.key.size() + record.value.size();
 
-  BlobIndex index;
-  BlobRecord record;
-  record.key = key;
-  record.value = value;
-  index.file_number = blob_handle_->GetNumber();
-  blob_builder_->Add(record, &index.blob_handle);
-  RecordTick(statistics(stats_), TITAN_BLOB_FILE_BYTES_WRITTEN,
-             index.blob_handle.size);
-  bytes_written_ += record.size();
-  if (ok()) {
-    index.EncodeTo(index_value);
-    if (blob_handle_->GetFile()->GetFileSize() >=
-        cf_options_.blob_file_target_size) {
-      FinishBlobFile();
+  std::unique_ptr<BlobFileBuilder::BlobRecordContext> ctx(
+      new BlobFileBuilder::BlobRecordContext);
+  AppendInternalKey(&ctx->key, ikey);
+  ctx->new_blob_index.file_number = blob_handle_->GetNumber();
+  blob_builder_->Add(record, std::move(ctx), contexts);
+}
+
+void TitanTableBuilder::AddToBaseTable(
+    const BlobFileBuilder::OutContexts& contexts, bool finishing) {
+  if (contexts.empty()) return;
+  for (const std::unique_ptr<BlobFileBuilder::BlobRecordContext>& ctx :
+       contexts) {
+    ParsedInternalKey ikey;
+    if (!ParseInternalKey(ctx->key, &ikey)) {
+      status_ = Status::Corruption(Slice());
+      return;
+    }
+    if (ctx->cached_data.is_cached) {
+      // write directly to base table
+      base_builder_->Add(ctx->key, ctx->cached_data.value);
+    } else {
+      RecordTick(statistics(stats_), TITAN_BLOB_FILE_BYTES_WRITTEN,
+                 ctx->new_blob_index.blob_handle.size);
+      bytes_written_ += ctx->new_blob_index.blob_handle.size;
+      if (ok()) {
+        std::string index_value;
+        ctx->new_blob_index.EncodeTo(&index_value);
+
+        ikey.type = kTypeBlobIndex;
+        std::string index_key;
+        AppendInternalKey(&index_key, ikey);
+        base_builder_->Add(index_key, index_value);
+      }
     }
   }
+  if (!finishing && blob_handle_->GetFile()->GetFileSize() >=
```

Comment:
We should have this check and call `FinishBlobFile` inside `AddBlob`, before calling `AddToBaseTable`.